File: node.js/ios-basic.js
'use strict';

var wd = require('wd');
var request = require('request');
var Promise = require('bluebird'); // jshint ignore:line
var fs = require('fs');

var filepath = 'outputs/';
fs.mkdir(filepath, function(ignored) {});

var apiToken = 'xyz'; // your API token from https://appetize.io/docs#request-api-token
var deviceType = 'iphone5s'; // iphone4s, iphone5s, iphone6s, iphone6splus, iphone7, iphone7plus, ipadair2
var publicKey = 'p7cc48c1k8pr1qvnz6r3quu098'; // replace with your own publicKey after uploading through website or API
var osVersion = '10.2'; // also supports 10.1, 10.0
var proxy = 'intercept'; // false for no proxy, or specify your own with http://proxy-example.com:port
var params = {
    hello: 'world',
    userId: 123
}; // optional data to pass to app at launch

var driver = wd.remote('https://' + apiToken + '@appium.appetize.io/wd/hub', 'promiseChain');

console.log('starting session');
driver.init({
    device: deviceType,
    publicKey: publicKey,
    osVersion: osVersion,
    proxy: proxy,
    params: params,
    automationName: 'XCUITest'
}).delay(5000)
.then(takeScreenshot)
.then(function() {
    console.log('tapping element');
    return driver.elementByXPath('//XCUIElementTypeButton[@name="Saved"]').click().delay(2000);
})
.then(takeScreenshot)
.then(function() {
    console.log('ending session');
    return driver.quit();
})
.then(function() {
    var timeStamp = Date.now();

    if (proxy == 'intercept') {
        downloadFile('https://api.appetize.io/v1/networkCapture/appiumId/' + driver.sessionID,
            filepath + 'appetize-' + deviceType + '-' + osVersion + '-' + publicKey + '-har-' + timeStamp + '.har',
            'networkCapture');
    }

    downloadFile('https://api.appetize.io/v1/debugLog/appiumId/' + driver.sessionID,
        filepath + 'appetize-' + deviceType + '-' + osVersion + '-' + publicKey + '-debugLog-' + timeStamp + '.txt',
        'debugLog');

    downloadFile('https://api.appetize.io/v1/screenRecording/appiumId/' + driver.sessionID,
        filepath + 'appetize-' + deviceType + '-' + osVersion + '-' + publicKey + '-screenRecording-' + timeStamp + '.mp4',
        'screenRecording');
})
.catch(function(error) {
    console.log(error);
});

function takeScreenshot() {
    console.log('take screenshot');
    return driver.takeScreenshot()
    .then(function(data) {
            // write file
            var filename = 'appetize-' + deviceType + '-' + osVersion + '-' + publicKey + '-screenshot-' + Date.now() + '.png';
            console.log('writing file to ' + filename);
            fs.writeFileSync(filepath + filename, data, 'base64');
        });
}

function downloadFile(url, filename, type) {
    console.log('Downloading ' + type + ' file: ' + url);

    request({
        url: url,
        encoding: type === 'screenRecording' ? null : undefined
    }, function(err, response, body) {
        if (err) throw err;

        if (response.statusCode !== 200) {
            console.log('File download failed with ' + response.statusCode);
            return;
        }

        if (type === 'networkCapture') {
            var har = JSON.parse(body);
            console.log('HAR has ' + har.log.entries.length + ' entries');
        }
        else if (type === 'debugLog') {
            console.log('Debug log has ' + body.split('\n').length + ' lines');
        }

        console.log('Writing ' + type + ' to ' + filename);
        fs.writeFileSync(filename, body);
    });
}

File: java/src/main/java/IosBasic.java
import com.google.common.io.Files;
import io.appium.java_client.AppiumDriver;
import io.appium.java_client.android.AndroidDriver;
import org.openqa.selenium.OutputType;
import org.openqa.selenium.remote.DesiredCapabilities;
import org.json.simple.JSONObject;

import java.io.File;
import java.io.FileOutputStream;
import java.io.IOException;
import java.net.URL;
import java.nio.channels.Channels;
import java.nio.channels.ReadableByteChannel;

public class IosBasic {
    String apiToken = "xyz"; // your API token from https://appetize.io/docs#request-api-token
    String deviceType = "iphone5s"; // iphone5s, iphone6s, iphone6splus, iphone7, iphone7plus, ipadair2
    String publicKey = "p7cc48c1k8pr1qvnz6r3quu098"; // replace with your own publicKey after uploading through website or API
    String osVersion = "10.3"; // also supports 10.2, 10.1, 10.0
    String proxy = "intercept"; // false for no proxy, or specify your own with http://proxy-example.com:port
    String outputDirectory = "./outputs/";

    public void run() throws Exception {
        new File(outputDirectory).mkdir();

        JSONObject params = new JSONObject(); // will be passed to app at launch
        params.put("hello", "world");
        params.put("userId", 123);

        DesiredCapabilities capabilities = new DesiredCapabilities();
        capabilities.setCapability("device", deviceType);
        capabilities.setCapability("publicKey", publicKey);
        capabilities.setCapability("osVersion", osVersion);
        capabilities.setCapability("proxy", proxy);
        capabilities.setCapability("params", params.toString());

        URL remote = new URL("https://" + apiToken + "@appium.appetize.io/wd/hub");
        AppiumDriver driver = new AndroidDriver(remote, capabilities);

        System.out.println("session started");

        Thread.sleep(5000);
        takeScreenshot(driver);

        System.out.println("tapping element");
        driver.findElementByXPath("//XCUIElementTypeButton[@name=\"Saved\"]").click();

        Thread.sleep(2000);
        takeScreenshot(driver);

        System.out.println("ending session");
        String sessionId = driver.getSessionId().toString();
        driver.quit();

        // download files
        String timeStamp = "" + System.currentTimeMillis();
        if ("intercept".equals(proxy)) {
            downloadFile("https://api.appetize.io/v1/networkCapture/appiumId/" + sessionId,
                    "appetize-" + deviceType + "-" + osVersion + "-" + publicKey +
                    "-har-" + timeStamp + ".har");
        }

        downloadFile("https://api.appetize.io/v1/debugLog/appiumId/" + sessionId,
                "appetize-" + deviceType + "-" + osVersion + "-" + publicKey +
                        "-debugLog-" + timeStamp + ".txt");

        downloadFile("https://api.appetize.io/v1/screenRecording/appiumId/" + sessionId,
                "appetize-" + deviceType + "-" + osVersion + "-" + publicKey +
                        "-screenRecording-" + timeStamp + ".mp4");
    }

    void takeScreenshot(AppiumDriver driver) throws IOException {
        System.out.println("take screenshot");
        String fileName = "appetize-" + deviceType + "-" + osVersion + "-" + publicKey +
                "-screenshot-" + System.currentTimeMillis() + ".png";
        File destination = new File(outputDirectory, fileName);

        File screenShot = driver.getScreenshotAs(OutputType.FILE);
        Files.move(screenShot, destination);
    }

    void downloadFile(String url, String fileName) throws Exception {
        System.out.println("downloading " + url);
        URL remoteUrl = new URL(url);
        ReadableByteChannel rbc = Channels.newChannel(remoteUrl.openStream());
        FileOutputStream fos = new FileOutputStream(new File(outputDirectory, fileName));
        fos.getChannel().transferFrom(rbc, 0, Long.MAX_VALUE);
    }
}